### DISCUSSION TITLE
Be able to run PHP code at the paused position

### DISCUSSION THREAD

**Discussion by marnickmenting** [spatie/laravel-ray#160] [Ideas]
Coming from Xdebug, I often use the Console to execute PHP commands _in the context of the breakpoint_. This is particular helpful to check if some Service if loaded or not or check the result of some helper function before and after running some code.

Not sure how you can do this exactly, as obviously Ray is not integrated in PHPStorm.

![](https://blog.jetbrains.com/wp-content/uploads/2015/09/phpstorm-InteractiveConsole1.png)

**Comment by freekmurze** [comment]
It's technically a tall orde. If we figure out to do this, we'll surely add it.

**Reply by marnickmenting** [reply]
I'd imagine. Was worth a shot 😉

**Reply by tpetry** [reply]
@freekmurze How about changing the lock so the release of the log is requested from the ray application (instead of pusing data to the port it's now requesting information). And when the lock is not released the ray application could define some php code to which is fed to `eval()` and the return value again `ray()`ed.

**Reply by tpetry** [reply]
My idea yesterday was not really working because of a scope problem, but hey, i found another solution.

So the best we can currently do is something like:
```php
ray(get_defined_vars())->pause();
```

But this will only stop the script and not output the values. I don't know whether this is is expected or an bug. But we could do something more intelligently with this idea:

```php
ray()->pause(get_defined_vars(), $this);
```

So with this approach we would pass the actual scope to the paused execution context. As the ray library is now aware of the current variables my old idea from yesterday is working. Similar to artisan tinker console, typed php code within the ray desktop could now be executed __within__ the context of the current running request __and__ you would have access to all variables of the current scope the same as when the code was originally written there. By using [bind](https://www.php.net/manual/en/closure.bind.php) we can eben change the `$this` reference so your typed php code can access private/protected methods, attributes.

@freekmurze What's your opinion on this? It would be a huuuge leap forward for ray taking it almost to the same functionality of a full debugger.

**Reply by marnickmenting** [reply]
I didn't know about `get_defined_vars()`, thanks for that. `ray(get_defined_vars())` already gives some useful information.
